zebrad/src/components/sync/tests/vectors.rs

```diff
@@ -0,0 +1,128 @@
+       "Block verify should allow for a full checkpoint download, a sync restart, then a full checkpoint re-download"
+    );
+
+    // This constraint avoids spurious failures after checkpointing has finished
+    assert!(
+        BLOCK_VERIFY_TIMEOUT.as_secs()
+            > 2 * zebra_chain::parameters::NetworkUpgrade::Blossom
+                .target_spacing()
+                .num_seconds() as u64,
+        "Block verify should allow for at least one new block to be generated and distributed"
+    );
+}
+
+/// Test that calls to [`ChainSync::request_genesis`] are rate limited.
+#[test]
+fn request_genesis_is_rate_limited() {
+    // create some counters that will be updated inside async blocks
+    let peer_requests_counter = Arc::new(AtomicU8::new(0));
+    let peer_requests_counter_in_service = Arc::clone(&peer_requests_counter);
+    let state_requests_counter = Arc::new(AtomicU8::new(0));
+    let state_requests_counter_in_service = Arc::clone(&state_requests_counter);
+
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    // create a fake peer service that respond with `Error` to `BlocksByHash` or
+    // panic in any other type of request.
+    let peer_service = tower::service_fn(move |request| {
+        match request {
+            zebra_network::Request::BlocksByHash(_) => {
+                // Track the call
+                peer_requests_counter_in_service.fetch_add(1, Ordering::SeqCst);
+                // Respond with `Error`
+                future::err("block not found".into())
+            }
+            _ => unreachable!("no other request is allowed"),
+        }
+    });
+
+    // create a state service that respond with `None` to `Depth` or
+    // panic in any other type of request.
+    let state_service = tower::service_fn(move |request| {
+        match request {
+            zebra_state::Request::Depth(_) => {
+                // Track the call
+                state_requests_counter_in_service.fetch_add(1, Ordering::SeqCst);
+                // Respond with `None`
+                future::ok(zebra_state::Response::Depth(None))
+            }
+            _ => unreachable!("no other request is allowed"),
+        }
+    });
+
+    // create a verifier service that will always panic as it will never be called
+    let verifier_service =
+        tower::service_fn(
+            move |_| async move { unreachable!("no request to this service is allowed") },
+        );
+
+    // start the sync
+    let mut chain_sync = ChainSync::new(
+        &ZebradConfig::default(),
+        peer_service,
+        state_service,
+        verifier_service,
+    );
+
+    // run `request_genesis()` with a timeout of 13 seconds
+    runtime.block_on(async move {
+        let _ = timeout(Duration::from_secs(13), chain_sync.request_genesis()).await;
```

Comment:
Here's how to calculate the timeout you need:
```suggestion
        // allow extra wall clock time for tests on CPU-bound machines
        let retries_timeout = (RETRIES_TO_RUN - 1) * GENESIS_TIMEOUT_RETRY + GENESIS_TIMEOUT_RETRY/2;
        let _ = timeout(Duration::from_secs(retries_timeout), chain_sync.request_genesis()).await;
```